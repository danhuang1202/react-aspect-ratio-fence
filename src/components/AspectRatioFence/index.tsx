import { createElement, FunctionComponent, ReactNode } from 'react'
import cns from 'classnames'

// @ts-ignore
import styles from './style.css'

interface Props {
  /**
   *  html tag name
   *  @default div
   * */
  elementType?: string
  /**
   *  The aspect ratio of an image describes the proportional
   *  relationship between width and height.
   *   @default 1
   **/
  ratio: number
  /** Custom class name */
  customClass?: string
  /** custom children node */
  children?: ReactNode
}

const AspectRatioFence: FunctionComponent<Props> = ({
  elementType = 'div',
  ratio = 1,
  customClass,
  children
}) =>
  createElement(elementType, {
    style: { '--ratio': `(${ratio})` },
    className: cns(styles.wrap, customClass),
    children
  })

export default AspectRatioFence
