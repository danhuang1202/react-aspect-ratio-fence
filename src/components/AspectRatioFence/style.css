.wrap {
  position: relative;
  width: 100%;
}

.wrap:before {
  content: '';
  display: block;
  width: 100%;
  height: 0;
  padding-bottom: calc(100% / (var(--ratio)));
}

.wrap > :first-of-type {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}
