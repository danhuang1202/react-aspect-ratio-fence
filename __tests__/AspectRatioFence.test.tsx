import React from 'react'
import renderer from 'react-test-renderer'
import AspectRatioFence from '@components/AspectRatioFence'

describe('[snapshot] AspectRatioFence', () => {
  it('AspectRatioFence', () => {
    const props = {
      ratio: 16 / 9
    }

    const dom = renderer.create(
      <AspectRatioFence {...props}>
        <img alt="AspectRatioFence" src="AspectRatioFence.png" />
      </AspectRatioFence>
    )
    expect(dom).toMatchSnapshot()
  })

  it('AspectRatioFence with custom elementType', () => {
    const props = {
      elementType: 'section',
      ratio: 4 / 3
    }

    const dom = renderer.create(
      <AspectRatioFence {...props}>
        <img alt="AspectRatioFence" src="AspectRatioFence.png" />
      </AspectRatioFence>
    )
    expect(dom).toMatchSnapshot()
  })
})
