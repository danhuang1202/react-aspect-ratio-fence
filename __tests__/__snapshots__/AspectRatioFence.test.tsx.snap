// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[snapshot] AspectRatioFence AspectRatioFence 1`] = `
<div
  className="AspectRatioFence_wrap_2HH2L"
  style={
    Object {
      "--ratio": "(1.7777777777777777)",
    }
  }
>
  <img
    alt="AspectRatioFence"
    src="AspectRatioFence.png"
  />
</div>
`;

exports[`[snapshot] AspectRatioFence AspectRatioFence with custom elementType 1`] = `
<section
  className="AspectRatioFence_wrap_2HH2L"
  style={
    Object {
      "--ratio": "(1.3333333333333333)",
    }
  }
>
  <img
    alt="AspectRatioFence"
    src="AspectRatioFence.png"
  />
</section>
`;
